### ISSUE TITLE
verifyConditions step fails in subgroup projects

### ISSUE THREAD

**Issue by p-ob** [semantic-release/gitlab#19]
My organization has a private GitLab instance where it's quite common to organize projects into group/subgroup relationships. I did find issue #2, but that seems to be strictly related to the usage of `/` in project names and properly encoding that for the API url.

I've hunted down the cause of this issue, and I believe it lies in the odd behavior of permission inheritance in GitLab. When making the API call to fetch the project with the access token provided, if the associated user has push privileges on a containing group level, but isn't added as an explicit member to the project, the API responds with the following permissions:
```json
...
    "permissions": {
        "project_access": null,
        "group_access": null
    }
...
```

# Repro:

1. Create a group in GitLab (e.g. gitlab.com/foo)
1. Create a subgroup under the previous group (e.g. gitlab.com/foo/bar)
1. Create a project under the subgroup (e.g. gitlab.com/foo/bar/baz)
1. Add a user with permissions to push (e.g. Master) to the "foo" group
1. Generate an access token for that user with read_user, api permissions
1. Add that token to baz (can also be added at the group levels for either foo or bar) as GL_TOKEN
1. Commit to baz with a .gitlab-ci.yml file configured to call semantic-release with @semantic-release/gitlab 

# Expected Behavior
The build and release succeeds

# Actual Behavior:
The build fails with `EGHNOPERMISSION` at the verifyConditions step of @semantic-release/gitlab

# Workaround
A bit annoying, but you can assign the user explicitly to each project you want to use semantic-release/gitlab with.

**Comment by p-ob** [comment]
Well, I did look into how to handle the "what if you have access at a different level?" The fix I was thinking of would be to try to "happy path" as it exists today, and if that fails (e.g. returns null for both `project_access` and `group_access`): 
1. check if the project is part of a subgroup (which looks to exist on the API call made to fetch the project under the "namespace.parent_id", which points to a the group that contains the subgroup)
1. if no, fail. if yes, get the user for the token at api/v4/user to find the user's id
1. check if the user is a member of the parent_id group with appropriate permissions

The only downside is that I'm not sure of the limit (if any) of how many subgroups you can add to a particular group.

It definitely does feel like a shortcoming of the GitLab api (deserving of an enhancement request), but because GitLab does technically support this concept, it seems to be a bug in this project with the current GitLab API.

**Comment by pvdlg** [comment]
Thanks for the analysis!
All that seems quite complex and require deep knowledge of GitLab specificities/weirdness. So I'm not sure we want to implement and maintain in this plugin. We would like to keep this plugin specific to semantic-release.

What I would suggest is to open a feature request with GitLab, proposing an API improvement. Ideally we should be able to make an API call passing a token and a project id and know right away what permission we have. An API user shouldn't have to reverse engineer the GitLab internal mechanism to know if a user has permission on a repo. In addition making an API for each subgroup until we find one that give us permission, is really not efficient. Going up the tree should be done by GitLab on server side. 

An other solution, in case GitLab refuse to implement such feature would be to implement the algorithm you describe in its own module. Then we could use this module in this plugin.
That would also be beneficial for the community as a whole, as I'm sure we're not the first one encountering this issue.

**Comment by p-ob** [comment]
Good idea. When I have some free time, I'll see what I can whip up.

For reference, looks like I'm not the only one annoyed by the limitations of the GitLab API: https://gitlab.com/gitlab-org/gitlab-ce/issues/43684

**Comment by pvdlg** [comment]
Thanks. The GitLab issue has been opened a week ago. Let's see if it get some traction.
Having it implemented would be the ideal solution.

**Comment by kopax** [comment]
Hi, thanks a lot for that analysis. We are hitting the same error `EGHNOPERMISSION`.

We have been trying to use our `GL_TOKEN` to push (outside semantic-release) within the gitlab-ci and we ended up using an `SSH_PRIVATE_KEY` and ssh protocol.

I have one question @pvdlg , are you using the `GL_TOKEN` (user access token) tu push to the repo? I figured that it was not possible at all. This can be related.

It seems that semantic-release read the repository url from the `package.json` and use that one, but the GL_TOKEN is still required.

We concluded the push was only possible through ssh, so we can work with semantic release only with ssh.

**Comment by trumbitta** [comment]
@kopax did your solution with ssh work with `@semantic-release/gitlab`? And how did you add the private key to Gitlab?

**Comment by trumbitta** [comment]
A note: to make it finally work with `GL_TOKEN` I also had to use the optional `GL_URL` and `GL_PREFIX` variables instead of the configuration in `package.json` that wasn't working for me.

**Comment by joshdcomp** [comment]
Running into this issue as well, if Gitlab subgroups aren't supported with `verifyConditions`, can a note get added to the documentation saying as much? I spent a whole day and a good portion of this morning debugging our entire GL environment before finding this thread

And per @trumbitta's note, I didn't have the same luck—I ended up having to set `verifyConditions: []` in my `.releaserc` file because omitting it completely invokes Github

**Comment by ClemMakesApps** [comment]
I also ran into this issue with https://gitlab.com/gitlab-org/frontend/gitlab-ui/
The user that was setup to run `semantic-release` already had `push` access from the parent group level but it was still failing as replicated in this issue. I had to explicitly add the user to my subgroup project as a `maintainer` (push access) before it started working again. 

Also, I work at GitLab and just pinged one of our product managers to take a look at https://gitlab.com/gitlab-org/gitlab-ce/issues/43684 Hopefully we can get some traction there

**Comment by kopax** [comment]
@ClemMakesApps hi, if you work at GitLab, maybe you can help us with an issue we have between Gitlab-CI, semantic-release and `@semantic-release/gitlab`:

The gitlab plugin push to gitlab, we are deploying the documentation avec the semantic-release but because of that push and the current gitlab page driver, the page documentation was failing, it is reported in details here:

https://gitlab.com/gitlab-org/gitlab-ce/issues/46689

The only solution using `@semantic-release/gitlab` to keep deploying page was to trigger the build on the tagged version using an api call and passing all the secrets, which now is a security concern because of that issue:

https://gitlab.com/gitlab-org/gitlab-ce/issues/20422

Both aren't hard to solve, I even offered to help but the Gitlab doesn't seems to worry about having semantic-release work with their page nor to give a way to hide secure variables passed to a trigger.

Thanks in advance.

**Comment by eReGeBe** [comment]
Update on https://gitlab.com/gitlab-org/gitlab-ce/issues/43684: we [merged a change](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/24403) that will return the maximum group access level from across the subgroup hierarchy, which should solve the problem reported on this issue. It'll be included on the next GitLab release (11.8).

**Comment by pvdlg** [comment]
Thanks. I guess this issue can be closed then.
